public/app/core/history/richHistoryLocalStorage.ts

```diff
@@ -0,0 +1,172 @@
+import RichHistoryStorage, {
+  RichHistoryServiceError,
+  RichHistoryStorageWarning,
+  RichHistoryStorageWarningDetails,
+} from './richHistoryStorage';
+import { RichHistoryQuery } from '../../types';
+import store from '../store';
+import { DataQuery } from '@grafana/data';
+import { isEqual, omit } from 'lodash';
+import { createRetentionPeriodBoundary, RICH_HISTORY_SETTING_KEYS } from './richHistoryLocalStorageUtils';
+
+export const RICH_HISTORY_KEY = 'grafana.explore.richHistory';
+export const MAX_HISTORY_ITEMS = 10000;
+
+/**
+ * Local storage implementation for Rich History. It keeps all entries in browser's local storage.
+ */
+export default class RichHistoryLocalStorage implements RichHistoryStorage {
+  /**
+   * Return all history entries, perform migration and clean up entries not matching retention policy.
+   */
+  async getRichHistory(): Promise<RichHistoryQuery[]> {
+    const richHistory: RichHistoryQuery[] = store.getObject(RICH_HISTORY_KEY, []);
+    const transformedRichHistory = migrateRichHistory(richHistory);
+    return transformedRichHistory;
+  }
+
+  async addToRichHistory(richHistoryQuery: RichHistoryQuery): Promise<RichHistoryStorageWarningDetails | undefined> {
+    let richHistory = cleanUp(await this.getRichHistory());
+
+    /* Compare queries of a new query and last saved queries. If they are the same, (except selected properties,
+     * which can be different) don't save it in rich history.
+     */
+    const newQueriesToCompare = richHistoryQuery.queries.map((q) => omit(q, ['key', 'refId']));
+    const lastQueriesToCompare =
+      richHistory.length > 0 &&
+      richHistory[0].queries.map((q) => {
+        return omit(q, ['key', 'refId']);
+      });
+
+    if (isEqual(newQueriesToCompare, lastQueriesToCompare)) {
+      const error = new Error('Entry already exists');
+      error.name = 'DuplicatedEntry';
```

Comment:
Should we use RichHistoryServiceError.DuplicatedEntry here?
```suggestion
      error.name = RichHistoryServiceError.DuplicatedEntry;
```

Reply:
Fixed in 2fa591531af83350ecda57c0d9163ed342458f9b